File: mercadolivre/src/main/java/com/jornadadev/mercadolivre/config/ExistsIdValidator.java
package com.jornadadev.mercadolivre.config;

import lombok.RequiredArgsConstructor;

import javax.persistence.EntityManager;
import javax.persistence.Query;
import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import java.util.List;

@RequiredArgsConstructor
public class ExistsIdValidator implements ConstraintValidator<ExistsId, Object> {

    private final EntityManager em;
    private String domainAttribute;
    private Class<?> aClass;

    @Override
    public void initialize(ExistsId constraintAnnotation) {
        domainAttribute = constraintAnnotation.fieldName();
        aClass = constraintAnnotation.domainClass();
    }

    @Override
    public boolean isValid(Object valor, ConstraintValidatorContext constraintValidatorContext) {
        if (valor == null)
            return true;

        Query query = em.createQuery("select 1 from " + aClass.getName() + " where " +
                                     domainAttribute + " =:value ");
        query.setParameter("value", valor);
        List<?> list = query.getResultList();
        return !list.isEmpty();
    }
}

File: yfood/src/main/kotlin/com/jornadadev/yfood/config/ExistsIdValidator.kt
package com.jornadadev.yfood.config

import javax.persistence.EntityManager
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext
import kotlin.reflect.KClass

class ExistsIdValidator (private val em: EntityManager) : ConstraintValidator<ExistsId, Any?> {
    private lateinit var domainAttribute: String
    private lateinit var aClass: KClass<*>
    override fun initialize(constraintAnnotation: ExistsId) {
        domainAttribute = constraintAnnotation.fieldName
        aClass = constraintAnnotation.domainClass
    }

    override fun isValid(valor: Any?, constraintValidatorContext: ConstraintValidatorContext): Boolean {
        if (valor == null) return true
        val query = em.createQuery("select 1 from " + aClass.simpleName + " where " +
                domainAttribute + " =:value ")
        query.setParameter("value", valor)
        val list = query.resultList
        return list.isNotEmpty()
    }
}